File: trunk/Java/OPSJLib/src/ops/RequestReply.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package ops;

import ops.OPSObject;
import ops.Publisher;
import ops.Subscriber;
import ops.Topic;

/**
 *
 * @author Anton
 */
public class RequestReply<ReqType extends Request, RepType extends Reply>
{

    long timeout;
    int nrResends = 1;
    Publisher publisher;
    Subscriber subscriber;
    RepType reply;
    static int reqInt = 0;
    private String key;

    public long getTimeout()
    {
        return timeout;
    }

    public void setTimeout(long timeout)
    {
        this.timeout = timeout;
    }

    public int getNrResends()
    {
        return nrResends;
    }

    public void setNrResends(int nrResends)
    {
        this.nrResends = nrResends;
    }

    public RequestReply(Topic reqTopic, Topic repTopic, String key) throws CommException
    {
        publisher = new Publisher(reqTopic);
        subscriber = new Subscriber(repTopic);
        this.key = key;
        subscriber.addFilterQoSPolicy(new KeyFilterQoSPolicy(key));
        subscriber.start();
    }

    public RepType request(final ReqType request) throws Exception
    {
        reqInt ++;
        request.requestId = key + reqInt;
        reply = null;
        Thread t = new Thread(new Runnable()
        {
            public void run()
            {
                reply = (RepType)subscriber.waitForNextData(timeout);
                if(!reply.requestId.equals(request.requestId))
                {
                    reply = null;
                }
            }
        });
        t.start();


        for (int i = 0; i < nrResends; i++)
        {
            publisher.writeAsOPSObject(request);
            try
            {
                t.join(timeout / nrResends);
                break;
            }
            catch(InterruptedException ex){}
        }
 
        //Reply will be null if we don't get a reply within timeout.
        return reply;
    }
}

File: C#/Ops/RequestReply.cs
///////////////////////////////////////////////////////////
//  RequestReply.cs
//  Implementation of the Class RequestReply
//  Created on:      12-nov-2011 09:25:35
//  Author:
///////////////////////////////////////////////////////////

using System.Threading;

namespace Ops 
{
	public class RequestReply<ReqType, RepType> where ReqType : Request where RepType : Reply
    {
		private string key;
		internal int nrResends = 1;
		internal Publisher publisher;
		internal RepType reply;
		internal static int reqInt = 0;
		internal Subscriber subscriber;
		internal long timeout;
        internal string lastRequestId;

        public long GetTimeout()
        {
            return timeout;
        }

        public void SetTimeout(long timeout)
        {
            this.timeout = timeout;
        }

        public int GetNrResends()
        {
            return nrResends;
        }

        public void SetNrResends(int nrResends)
        {
            this.nrResends = nrResends;
        }

        public RequestReply(Topic reqTopic, Topic repTopic, string key) 
        {
            publisher = new Publisher(reqTopic);
            subscriber = new Subscriber(repTopic);
            this.key = key;
            subscriber.AddFilterQoSPolicy(new KeyFilterQoSPolicy(key));
            subscriber.Start();
        }

        public RepType Request(ReqType request)
        {
            reqInt ++;
            request.requestId = key + reqInt;
            lastRequestId = request.requestId;
            reply = null;

            Thread thread = new Thread(new ThreadStart(Run));
            thread.IsBackground = true;
            thread.Start();

            for (int i = 0; i < nrResends; i++)
            {
                publisher.WriteAsOPSObject(request);
                try
                {
                    thread.Join((int)(timeout / nrResends));
                    break;
                }
                catch (ThreadAbortException)
                {
                }
            }
     
            // Reply will be null if we don't get a reply within timeout.
            return reply;
        }

        public void Run()
        {
            reply = (RepType)subscriber.WaitForNextOpsObjectData(timeout);
            if (!reply.requestId.Equals(lastRequestId))
            {
                reply = null;
            }
        }
	}

}